go/vt/vtgate/semantics/errors.go

```diff
@@ -25,183 +25,272 @@ import (
-	return &Error{
-		Code: code,
-		args: args,
-	}
-}
-
-var errors = map[ErrorCode]info{
-	UnionColumnsDoNotMatch: {
-		format: "The used SELECT statements have a different number of columns",
-		state:  vterrors.WrongNumberOfColumnsInSelect,
-		code:   vtrpcpb.Code_FAILED_PRECONDITION,
-	},
-	UnsupportedMultiTablesInUpdate: {
-		format: "multiple tables in update",
-		typ:    Unsupported,
-	},
-	TableNotUpdatable: {
-		format: "The target table %s of the UPDATE is not updatable",
-		state:  vterrors.NonUpdateableTable,
-		code:   vtrpcpb.Code_INVALID_ARGUMENT,
-	},
-	UnsupportedNaturalJoin: {
-		format: "%s",
-		typ:    Unsupported,
-	},
-	UnionWithSQLCalcFoundRows: {
-		format: "SQL_CALC_FOUND_ROWS not supported with union",
-		typ:    Unsupported,
-	},
-	SQLCalcFoundRowsUsage: {
-		format: "Incorrect usage/placement of 'SQL_CALC_FOUND_ROWS'",
-		code:   vtrpcpb.Code_INVALID_ARGUMENT,
-	},
-	CantUseOptionHere: {
-		format: "Incorrect usage/placement of '%s'",
-		state:  vterrors.CantUseOptionHere,
-		code:   vtrpcpb.Code_INVALID_ARGUMENT,
-	},
-	MissingInVSchema: {
-		format: "Table information is not provided in vschema",
-		code:   vtrpcpb.Code_INVALID_ARGUMENT,
-	},
-	NotSequenceTable: {
-		format: "NEXT used on a non-sequence table",
-		code:   vtrpcpb.Code_INVALID_ARGUMENT,
-	},
-	NextWithMultipleTables: {
-		format: "Next statement should not contain multiple tables",
-		typ:    Bug,
-	},
-	LockOnlyWithDual: {
-		format: "%v allowed only with dual",
-		code:   vtrpcpb.Code_UNIMPLEMENTED,
-	},
-	QualifiedOrderInUnion: {
-		format: "Table %s from one of the SELECTs cannot be used in global ORDER clause",
-	},
-	JSONTables: {
-		format: "json_table expressions",
-		typ:    Unsupported,
-	},
-	Buggy: {
-		format: "%s",
-		typ:    Bug,
-	},
-	ColumnNotFound: {
-		format: "symbol %s not found",
-		state:  vterrors.BadFieldError,
-		code:   vtrpcpb.Code_INVALID_ARGUMENT,
-	},
-	AmbiguousColumn: {
-		format: "Column '%s' in field list is ambiguous",
-		state:  vterrors.BadFieldError,
-		code:   vtrpcpb.Code_INVALID_ARGUMENT,
-	},
-}
-
-func (n *Error) Error() string {
-	f, ok := errors[n.Code]
-	if !ok {
-		return "unknown error"
-	}
-
-	format := f.format
-
-	if f.id != "" {
-		format = fmt.Sprintf("%s: %s", f.id, format)
-	}
+func printf(e error, msg string, args ...any) string {
+	format := msg
 
-	switch f.typ {
-	case Unsupported:
-		format = "VT12001: unsupported: " + format
-	case Bug:
-		format = "VT13001: [BUG] " + format
+	if eId, ok := e.(identifiableError); ok {
+		format = fmt.Sprintf("%s: %s", eId.id(), format)
 	}
 
-	var args []any
-	for _, arg := range n.args {
-		ast, isAST := arg.(sqlparser.SQLNode)
-		if isAST {
-			args = append(args, sqlparser.String(ast))
-		} else {
-			args = append(args, arg)
+	if eTyp, ok := e.(typedError); ok {
+		switch eTyp.typ() {
+		case UnsupportedErrorType:
+			format = "VT12001: unsupported: " + format
+		case BugErrorType:
+			format = "VT13001: [BUG] " + format
```

Comment:
How about creating `vterrors.VitessError`s here like:
```go
format = fmt.Sprintf(format, args...)
...
switch eTyp.typ() {
case UnsupportedErrorType:
  formatErr = vterrors.VT12001(format)
...
return formatErr.Error()
```

Reply:
I think this is fixed by @vmg 's https://github.com/vitessio/vitess/pull/12572/commits/4e1184a39d2bc286f45456808d88e37915dc5e07

Reply:
nah, it wasn't